app/jobs/repo_collaborators_job.rb

```diff
@@ -0,0 +1,61 @@
+class RepoCollaboratorsJob < CourseJob
+
+  @job_name = "Get All Repository Collaborators"
+  @job_short_name = "get_repo_contributors"
+
+  def perform(course_id)
+    ActiveRecord::Base.connection_pool.with_connection do
+      super
+      course = Course.find(course_id)
+      course_repos = course.github_repos
+      course_students = course.roster_students
+
+      total_users_matched = 0; total_repos_matched = 0
+      course_repos.each do |repo|
+        users_matched = users_for_repo(repo, course_students)
+        total_users_matched += users_matched
+        total_repos_matched += users_matched >= 1 ? 1 : 0  # Repo is only counted in summary if it was matched to >= 1 user
+      end
+      summary = "#{total_users_matched} collaborators found for #{total_repos_matched} repositories."
+      create_completed_job_record(summary, course_id) # TODO: Replace this when rebasing other PR changes
+    end
+  end
+
+  # Substring matching and collaborator matching. Collaborator matching:
+  # Matching by asking the GitHub API for all the direct collaborators on the repository
+  # "direct" affiliation means that they actually contribute to the repo, not simply have permissions to it by default.
+  def users_for_repo(repo_record, students)
+    # Statistics gathering for job summary
+    users_matched = 0
+
+    repo_name = repo_record.name.downcase
+    repo_collaborators_req = github_machine_user.collaborators(repo_record.full_name, affiliation: "direct")
+    if repo_collaborators_req.respond_to? :each
+      repo_collaborators = repo_collaborators_req.map { |collaborator| collaborator.login}
+    else
+      repo_collaborators = []
+    end
+    students.each do |student|
+      unless student.user.nil? or student.user.username.nil?
```

Comment:
Wait, really!?    

So, looping over all the students for each repo makes this O(s * r) where s is | students | and r is | repos|

I wonder if we could do better by simply parsing the repo name breaking it up where there are underscores.   The convention is that since hyphens are allowed in github usernames we require that underscores be used to separate the lab name from the two github usernames.

I'm seeing that this convention is pretty consistent across Diba, Richert, and my courses.    So I think we can depend on it.

Also: don't hate me---but I'm starting to wonder whether we need the substring convention at all now that we have the more meaningful and reliable collaborator API listing.  That's really the ground truth.

But, your table might not have been a waste of time.  It may be helpful to have the information about what "type" of collaborator each user is (admin, read, write, etc.).    We might want that in place of knowing whether the match was found by the API or by the substring method.

Reply:
Substring matching removed in 4d38df8